File: packages/php-sdk/src/Resources/AccessPasses.php
<?php

namespace Wusul\Resources;

use Wusul\HttpClient;

/**
 * AccessPasses resource for managing access passes
 */
class AccessPasses
{
    private HttpClient $http;

    /**
     * Creates a new AccessPasses resource instance
     *
     * @param HttpClient $httpClient The HTTP client
     */
    public function __construct(HttpClient $httpClient)
    {
        $this->http = $httpClient;
    }

    /**
     * Issue a new access pass
     *
     * @param array $params Parameters for issuing the access pass
     *                      - cardTemplateId (required): Card template ID
     *                      - fullName (required): Full name of the pass owner
     *                      - email: Email address
     *                      - phoneNumber: Phone number
     *                      - cardNumber (required): Card number
     *                      - startDate (required): ISO8601 timestamp
     *                      - expirationDate (required): ISO8601 timestamp
     *                      - classification: Employment classification
     *                      - title: Job title
     *                      - employeePhoto: Base64 encoded image
     *                      - employeeId: Employee ID
     *                      - tagId: Tag identifier (if enabled)
     *                      - siteCode: Site code for H10301 format
     *                      - fileData: Hexadecimal data for proprietary formats
     *                      - memberId: Loyalty/membership ID (hotel use case)
     *                      - membershipStatus: Membership status (hotel use case)
     *                      - isPassReadyToTransact: Pass ready for NFC (hotel use case)
     *                      - tileData: Hotel tile display data
     *                      - reservations: Hotel reservation data
     *                      - metadata: Custom JSON object
     * @return array The created access pass with installation URL
     * @throws \Exception
     */
    public function issue(array $params): array
    {
        return $this->http->post('/v1/access-passes', $params);
    }

    /**
     * List access passes with optional filtering
     *
     * @param array|null $params Optional parameters for filtering
     *                           - templateId: Filter by card template ID
     *                           - state: Filter by state (active, suspended, unlink, deleted)
     * @return array Array of access passes
     * @throws \Exception
     */
    public function list(?array $params = null): array
    {
        $sigPayload = [];

        if (isset($params['templateId'])) {
            $sigPayload['template_id'] = $params['templateId'];
        }
        if (isset($params['state'])) {
            $sigPayload['state'] = $params['state'];
        }

        return $this->http->get('/v1/access-passes', $sigPayload);
    }

    /**
     * Update an access pass
     *
     * @param array $params Parameters for updating the access pass
     *                      - accessPassId (required): The ID of the access pass
     *                      - fullName: Updated full name
     *                      - classification: Updated classification
     *                      - expirationDate: Updated expiration date
     *                      - employeePhoto: Updated employee photo
     *                      - title: Updated job title
     *                      - employeeId: Updated employee ID
     *                      - fileData: Updated file data (Android)
     *                      - isPassReadyToTransact: Updated NFC ready status
     *                      - tileData: Updated tile data
     *                      - reservations: Updated reservations
     *                      - metadata: Updated metadata
     * @return array The updated access pass
     * @throws \Exception
     */
    public function update(array $params): array
    {
        $accessPassId = $params['accessPassId'] ?? null;
        if (!$accessPassId) {
            throw new \InvalidArgumentException('accessPassId is required');
        }

        unset($params['accessPassId']);
        return $this->http->patch("/v1/access-passes/{$accessPassId}", $params);
    }

    /**
     * Suspend an access pass
     *
     * @param string $accessPassId The ID of the access pass to suspend
     * @return array Success response
     * @throws \Exception
     */
    public function suspend(string $accessPassId): array
    {
        return $this->http->post("/v1/access-passes/{$accessPassId}/suspend");
    }

    /**
     * Resume a suspended access pass
     *
     * @param string $accessPassId The ID of the access pass to resume
     * @return array Success response
     * @throws \Exception
     */
    public function resume(string $accessPassId): array
    {
        return $this->http->post("/v1/access-passes/{$accessPassId}/resume");
    }

    /**
     * Unlink an access pass from the user's device
     *
     * @param string $accessPassId The ID of the access pass to unlink
     * @return array Success response
     * @throws \Exception
     */
    public function unlink(string $accessPassId): array
    {
        return $this->http->post("/v1/access-passes/{$accessPassId}/unlink");
    }

    /**
     * Delete an access pass permanently
     *
     * @param string $accessPassId The ID of the access pass to delete
     * @return array Success response
     * @throws \Exception
     */
    public function delete(string $accessPassId): array
    {
        return $this->http->post("/v1/access-passes/{$accessPassId}/delete");
    }
}

File: packages/ruby-sdk/lib/wusul/resources/access_passes.rb
module Wusul
  module Resources
    # AccessPasses resource for managing access passes
    class AccessPasses
      # Initialize the AccessPasses resource
      #
      # @param http_client [HttpClient] The HTTP client instance
      def initialize(http_client)
        @http = http_client
      end

      # Issue a new access pass
      #
      # @param params [Hash] Parameters for issuing the access pass
      # @option params [String] :card_template_id Unique identifier for the card template to use
      # @option params [String] :employee_id Unique identifier for the employee
      # @option params [String, nil] :tag_id Tag identifier (7 bytes, 14 chars hex)
      # @option params [String, nil] :site_code Site code from H10301 format (< 255)
      # @option params [String] :card_number Card number from H10301 format (< 65,535)
      # @option params [String, nil] :file_data Up to 64 chars of hex data
      # @option params [String] :full_name Full name of the access pass owner
      # @option params [String, nil] :email Email address
      # @option params [String, nil] :phone_number Contact phone number
      # @option params [String, nil] :classification Employment classification
      # @option params [String] :start_date ISO8601 timestamp when active
      # @option params [String] :expiration_date ISO8601 timestamp when expires
      # @option params [String, nil] :employee_photo Base64 encoded image
      # @option params [String, nil] :title Job title or department
      # @option params [String, nil] :member_id Loyalty/membership ID
      # @option params [String, nil] :membership_status Membership status
      # @option params [Boolean, nil] :is_pass_ready_to_transact Pass ready for NFC
      # @option params [Hash, nil] :tile_data Data for hotel tile display
      # @option params [Hash, nil] :reservations Data for hotel reservation
      # @option params [Hash, nil] :metadata Custom JSON object for additional data
      # @return [Hash] The created access pass with installation URL
      def issue(params)
        @http.post('/v1/access-passes', params)
      end

      # List access passes with optional filtering
      #
      # @param params [Hash, nil] Optional parameters for filtering
      # @option params [String, nil] :template_id Filter by card template ID
      # @option params [String, nil] :state Filter by state (active, suspended, unlink, deleted)
      # @return [Array<Hash>] Array of access passes
      def list(params = nil)
        sig_payload = {}

        if params
          sig_payload[:template_id] = params[:template_id] if params[:template_id]
          sig_payload[:state] = params[:state] if params[:state]
        end

        @http.get('/v1/access-passes', sig_payload.empty? ? nil : sig_payload)
      end

      # Update an access pass
      #
      # @param params [Hash] Parameters for updating the access pass
      # @option params [String] :card_id Unique identifier of the access pass to update
      # @option params [String, nil] :employee_id Updated employee identifier
      # @option params [String, nil] :full_name Updated full name
      # @option params [String, nil] :classification Updated employment classification
      # @option params [String, nil] :expiration_date Updated expiration date
      # @option params [String, nil] :employee_photo Updated base64 encoded image
      # @option params [String, nil] :title Updated job title
      # @option params [String, nil] :file_data Updated hex data
      # @option params [Boolean, nil] :is_pass_ready_to_transact Pass ready for NFC
      # @option params [Hash, nil] :tile_data Updated hotel tile data
      # @option params [Hash, nil] :reservations Updated hotel reservation data
      # @option params [Hash, nil] :metadata Updated custom metadata
      # @return [Hash] The updated access pass
      def update(params)
        card_id = params.delete(:card_id)
        raise ArgumentError, 'card_id is required' unless card_id

        @http.patch("/v1/access-passes/#{card_id}", params)
      end

      # Suspend an access pass
      #
      # @param card_id [String] The ID of the access pass to suspend
      # @return [Hash] Success response
      def suspend(card_id)
        raise ArgumentError, 'card_id is required' unless card_id

        @http.post("/v1/access-passes/#{card_id}/suspend")
      end

      # Resume a suspended access pass
      #
      # @param card_id [String] The ID of the access pass to resume
      # @return [Hash] Success response
      def resume(card_id)
        raise ArgumentError, 'card_id is required' unless card_id

        @http.post("/v1/access-passes/#{card_id}/resume")
      end

      # Unlink an access pass from the user's device
      #
      # @param card_id [String] The ID of the access pass to unlink
      # @return [Hash] Success response
      def unlink(card_id)
        raise ArgumentError, 'card_id is required' unless card_id

        @http.post("/v1/access-passes/#{card_id}/unlink")
      end

      # Delete an access pass permanently
      #
      # @param card_id [String] The ID of the access pass to delete
      # @return [Hash] Success response
      def delete(card_id)
        raise ArgumentError, 'card_id is required' unless card_id

        @http.post("/v1/access-passes/#{card_id}/delete")
      end
    end
  end
end
